jerry-core/ecma/operations/ecma-proxy-object.c

```diff
@@ -897,9 +946,185 @@ ecma_collection_t *
+
+  /* 10. */
+  if (trap_result == NULL)
+  {
+    return trap_result;
+  }
+
+  /* 11. */
+  ecma_value_t extensible_target = ecma_builtin_object_object_is_extensible (target_obj_p);
+
+  /* 12. */
+  if (ECMA_IS_VALUE_ERROR (extensible_target))
+  {
+    ecma_collection_free (trap_result);
+    return NULL;
+  }
+
+  /* 13. */
+  ecma_collection_t *target_keys = ecma_op_object_get_property_names (target_obj_p, ECMA_LIST_SYMBOLS);
+
+  /* 14. */
+  if (target_keys == NULL)
+  {
+    ecma_collection_free (trap_result);
+    return target_keys;
+  }
+
+  /* 16. */
+  ecma_collection_t *target_configurable_keys = ecma_new_collection ();
+
+  /* 17. */
+  ecma_collection_t *target_non_configurable_keys = ecma_new_collection ();
+
+  ecma_collection_t *ret_value = NULL;
+
+  /* 18. */
+  for (uint32_t i = 0; i < target_keys->item_count; i++)
+  {
+    ecma_property_descriptor_t target_desc;
+
+    ecma_string_t *prop_name_p = ecma_get_prop_name_from_value (target_keys->buffer_p[i]);
+
+    ecma_value_t status = ecma_op_object_get_own_property_descriptor (target_obj_p,
+                                                                      prop_name_p,
+                                                                      &target_desc);
+
+    if (ECMA_IS_VALUE_ERROR (status))
+    {
+      ecma_collection_free (trap_result);
+      goto free_target_collections;
+    }
+
+    ecma_value_t prop_value = ecma_make_prop_name_value (prop_name_p);
+
+    if (ecma_is_value_true (status)
+        && !(target_desc.flags & ECMA_PROP_IS_CONFIGURABLE))
+    {
+      ecma_collection_push_back (target_non_configurable_keys, prop_value);
+    }
+    else
+    {
+      ecma_collection_push_back (target_configurable_keys, prop_value);
+    }
+
+    if (ecma_is_value_true (status))
+    {
+      ecma_free_property_descriptor (&target_desc);
+    }
+  }
+
+  /* 19. */
+  if (ecma_is_value_true (extensible_target) && target_non_configurable_keys->item_count == 0)
+  {
+    ret_value = trap_result;
+    goto free_target_collections;
+  }
+
+  /* 20. */
+  ecma_collection_t *unchecked_result_keys = ecma_new_collection ();
+
+  ecma_collection_append (unchecked_result_keys, trap_result->buffer_p, trap_result->item_count);
+
+  for (uint32_t i = 0; i < unchecked_result_keys->item_count; i++)
+  {
+    ecma_string_t *unchecked_prop_name = ecma_get_prop_name_from_value (unchecked_result_keys->buffer_p[i]);
+    ecma_ref_ecma_string (unchecked_prop_name);
+  }
+
+  uint32_t unchecked_prop_name_counter = 0;
+
+  /* 21. */
+  if (ECMA_IS_VALUE_ERROR (ecma_proxy_object_own_property_keys_helper (target_non_configurable_keys,
+                                                                       unchecked_result_keys,
+                                                                       &unchecked_prop_name_counter)))
+  {
+    ecma_raise_type_error (ECMA_ERR_MSG ("Trap result did not include all non-configurable keys."));
+    goto free_unchecked_result_keys;
```

Comment:
It would be really good if we can remove most of the goto-s from the code.

Reply:
I'm not sure that would be a good idea, since as I see each goto saves 3-4 collection cleanup calls.

Reply:
What if we extract the 20-25 steps to a method? ex.:
```c
bool
proxy_check (ecma_collection_t *trap_result,
             ecma_collection_t *target_non_configurable_keys,
             ecma_collection_t *target_configurable_keys)
{
  /* 20 - 25 */
  ecma_collection_t *unchecked_result_keys = ecma_new_collection ();
  ecma_collection_append (unchecked_result_keys, trap_result->buffer_p, trap_result->item_count);
  for (uint32_t i = 0; i < unchecked_result_keys->item_count; i++)
  {
    ecma_string_t *unchecked_prop_name = ecma_get_prop_name_from_value (unchecked_result_keys->buffer_p[i]);
    ecma_ref_ecma_string (unchecked_prop_name);
  }

  bool check_ok = false;
  uint32_t unchecked_prop_name_counter = 0;

  /* 21. */
  if (ECMA_IS_VALUE_ERROR (ecma_proxy_object_own_property_keys_helper (target_non_configurable_keys,
                                                                       unchecked_result_keys,
                                                                       &unchecked_prop_name_counter)))
  {
    ecma_raise_type_error (ECMA_ERR_MSG ("Trap result did not include all non-configurable keys."));
  }
  else if (ecma_is_value_true (extensible_target))  /* 22. */
  {
    check_ok = true;
  }
  else if (ECMA_IS_VALUE_ERROR (ecma_proxy_object_own_property_keys_helper (target_configurable_keys,
                                                                       unchecked_result_keys,
                                                                       &unchecked_prop_name_counter))) /* 23. */
  {
    ecma_raise_type_error (ECMA_ERR_MSG ("Trap result did not include all configurable keys."));
  }
  else if (unchecked_result_keys->item_count != unchecked_prop_name_counter) /* 24. */
  {
    ecma_raise_type_error (ECMA_ERR_MSG ("Trap returned extra keys but proxy target is non-extensible"));
  }
  else
  {
    /* 25. */
    check_ok = true;
  }

  ecma_collection_free (unchecked_result_keys);

  return check_ok;
}

/*....*/

  /* 19. */
  if (ecma_is_value_true (extensible_target) && target_non_configurable_keys->item_count == 0)
  {
    ret_value = trap_result;
  }
  else if (proxy_check(trap_result, target_non_configurable_keys, target_configurable_keys) == true)
  {
    ret_value = trap_result;
  }
  else
  {
    JERRY_ASSERT (ret_value == NULL);
    ecma_collection_free (trap_result);
  }

free_target_collections:
  ecma_collection_destroy (target_keys);
  ecma_collection_free (target_configurable_keys);
  ecma_collection_free (target_non_configurable_keys);

```